File: app/app/src/main/java/com/example/hakonsreader/views/ContentLink.java
package com.example.hakonsreader.views;

import android.content.Context;
import android.content.Intent;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.ViewGroup;
import android.widget.ScrollView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;

import com.example.hakonsreader.R;
import com.example.hakonsreader.activites.DispatcherActivity;
import com.example.hakonsreader.api.model.Image;
import com.example.hakonsreader.api.model.RedditPost;
import com.example.hakonsreader.databinding.ContentLinkBinding;
import com.squareup.picasso.Callback;
import com.squareup.picasso.Picasso;

import java.util.List;

/**
 * Class for post contents that are a link. This extends ScrollView so that if the user
 * selects a max size for posts that is very low this view can be scrolled to view it in its entirety
 */
public class ContentLink extends Content {
    private static final String TAG = "ContentLink";

    private final ContentLinkBinding binding;

    public ContentLink(@NonNull Context context) {
        this(context, null, 0, 0);
    }
    public ContentLink(@NonNull Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0, 0);
    }
    public ContentLink(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        this(context, attrs, defStyleAttr, 0);
    }
    public ContentLink(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        binding = ContentLinkBinding.inflate(LayoutInflater.from(context), this, true);

        binding.thumbnail.setOnClickListener(v -> this.openLink());
        binding.link.setOnClickListener(v -> this.openLink());
    }

    @Override
    protected void updateView() {
        // The previews will (I believe) never be above 1080p, and that should be fine for most devices
        // TODO although this will use more data, so it might be reasonable to add a data saving setting where
        //  this image quality is reduced

        List<Image> previews = redditPost.getPreviewImages();
        if (redditPost.isNsfw()) {
            List<Image> obfuscatedPreviews = redditPost.getObfuscatedPreviewImages();
            if (obfuscatedPreviews != null && !obfuscatedPreviews.isEmpty()) {
                previews = obfuscatedPreviews;
            }
        }

        if (!previews.isEmpty()) {
            Image preview = previews.get(previews.size() - 1);

            // If the image was deleted the response for getting the image will return a 404
            // We could add a callback and reduce the image height in onError, but it looks kind of weird
            // since it has to do the request which causes it suddenly jump, so if the image has been deleted
            // we just let the size be as it is
            if (!preview.getUrl().isEmpty()) {
                Picasso.get()
                        .load(preview.getUrl())
                        .into(binding.thumbnail);
            }
        } else {
            // No thumbnail, reduce the height so it doesn't take up too much unnecessary space
            ViewGroup.LayoutParams params = binding.thumbnail.getLayoutParams();
            params.height = (int) getResources().getDimension(R.dimen.contentLinkNoThumbnailSize);
            binding.thumbnail.setLayoutParams(params);
        }

        binding.link.setText(redditPost.getUrl());
    }

    /**
     * Dispatches the link to {@link DispatcherActivity}
     */
    private void openLink() {
        Intent intent = new Intent(getContext(), DispatcherActivity.class);
        intent.putExtra(DispatcherActivity.URL_KEY, redditPost.getUrl());
        getContext().startActivity(intent);
    }
}

File: app/app/src/main/java/com/example/hakonsreader/views/ContentLink.kt
package com.example.hakonsreader.views

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import androidx.core.content.ContextCompat
import androidx.core.graphics.drawable.toBitmap
import androidx.core.util.Pair
import androidx.viewbinding.ViewBinding
import com.example.hakonsreader.R
import com.example.hakonsreader.activities.DispatcherActivity
import com.example.hakonsreader.api.model.RedditPost
import com.example.hakonsreader.databinding.ContentLinkBinding
import com.example.hakonsreader.databinding.ContentLinkSimpleBinding
import com.example.hakonsreader.misc.Settings
import com.example.hakonsreader.misc.getAppIconFromUrl
import com.example.hakonsreader.misc.getImageVariantsForRedditPost
import com.example.hakonsreader.views.util.cache
import com.squareup.picasso.Picasso

/**
 * View for displaying links from a [RedditPost]. This will display different views depending on
 * [Settings.dataSavingEnabled]
 */
class ContentLink @JvmOverloads constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0
) : Content(context, attrs, defStyleAttr) {

    val binding: ViewBinding = getCorrectBinding()

    /**
     * Retrieves and inflates the correct binding to use for the layout based on the users
     * data saving setting
     */
    private fun getCorrectBinding() : ViewBinding {
        return if (Settings.dataSavingEnabled()) {
            ContentLinkSimpleBinding.inflate(LayoutInflater.from(context), this, true)
        } else {
            ContentLinkBinding.inflate(LayoutInflater.from(context), this, true)
        }
    }


    override fun updateView() {
        val icon = getAppIconFromUrl(context, redditPost.url)

        if (binding is ContentLinkBinding) {
            this.updateViewNormal(icon)
        } else {
            this.updateViewSimple(icon)
        }

        setOnClickListener { openLink() }
    }

    override fun getTransitionViews(): MutableList<Pair<View, String>> {
        return super.getTransitionViews().also {
            it.add(Pair(binding.root, binding.root.transitionName))
        }
    }

    override fun getBitmap(): Bitmap? {
        return when (binding) {
            is ContentLinkBinding -> binding.thumbnail.drawable?.toBitmap()
            is ContentLinkSimpleBinding -> binding.thumbnail.drawable?.toBitmap()
            else -> null
        }
    }

    /**
     * Updates the view for the normal
     */
    private fun updateViewNormal(icon: Drawable?) {
        binding as ContentLinkBinding

        val variants = getImageVariantsForRedditPost(redditPost)

        val url = when {
            redditPost.isNsfw -> variants.nsfw
            redditPost.isSpoiler -> variants.spoiler
            else -> variants.normal
        }

        when {
            bitmap != null -> {
                binding.thumbnail.setImageBitmap(bitmap)
            }

            url != null -> {
                Picasso.get()
                        .load(url)
                        .cache(cache)
                        .into(binding.thumbnail)
            }

            else -> {
                val params: ViewGroup.LayoutParams = binding.thumbnail.layoutParams
                params.height = resources.getDimension(R.dimen.contentLinkNoThumbnailSize).toInt()
                binding.thumbnail.layoutParams = params
            }
        }

        if (icon != null) {
            binding.linkImage.setImageDrawable(icon)
        } else {
            binding.linkImage.setImageResource(R.drawable.ic_baseline_link_24)
        }

        binding.link.text = redditPost.url
    }

    /**
     * Updates the view for a simple view. This view uses only the thumbnail for the post, which
     * will be a smaller image and use less data
     */
    private fun updateViewSimple(icon: Drawable?) {
        binding as ContentLinkSimpleBinding
        binding.thumbnail.scaleType = ImageView.ScaleType.CENTER_CROP

        val thumbnail = redditPost.thumbnail

        when {
            bitmap != null -> {
                binding.thumbnail.setImageBitmap(bitmap)
            }

            // If no thumbnail is given, reddit might give it as "default"
            thumbnail.isNotBlank() && thumbnail != "default" -> {
                Picasso.get()
                        .load(thumbnail)
                        .into(binding.thumbnail)
            }

            else -> {
                // No thumbnail, set default link symbol
                // This should only be centered, not cropped to fit (as this would stretch the icon)
                binding.thumbnail.scaleType = ImageView.ScaleType.CENTER
                binding.thumbnail.setImageDrawable(ContextCompat.getDrawable(context, R.drawable.ic_baseline_link_24))
            }
        }

        if (icon != null) {
            binding.linkImage.setImageDrawable(icon)
        } else {
            binding.linkImage.setImageResource(R.drawable.ic_baseline_link_24)
        }

        binding.link.text = redditPost.url
    }


    /**
     * Dispatches the link to [DispatcherActivity]
     */
    private fun openLink() {
        val intent = Intent(context, DispatcherActivity::class.java)
        intent.putExtra(DispatcherActivity.EXTRAS_URL_KEY, redditPost.url)
        context.startActivity(intent)
    }
}